core/include/initcall.h

```diff
@@ -24,6 +24,7 @@ typedef TEE_Result (*initcall_t)(void);
 #define service_init_late(fn)		__define_initcall(4, fn)
 #define driver_init(fn)			__define_initcall(5, fn)
 #define driver_init_late(fn)		__define_initcall(6, fn)
+#define early_boot_final(fn)		__define_initcall(7, fn)
```

Comment:
`early_init` and `early_init_late` already mention `early` referring to Core early boot, I fear here `early_boot_final` is misleading.  I suggest `final_init` instead.

Also, here, you can't rely on `__define_initcall()` since initcalls are called before other Core initialization.
I rather suggest you do as you initially propose: implement a specific `__define_finalcall` macro and have the registered handlers called from the end of `generic_boot_init_primary()`.

Reply:
There isn't that much done after `call_initcalls()`. Perhaps we can convert `init_teecore()` to use only initcalls and even move the `release_external_dt()` into init calls? It doesn't have to be done in the PR though. If this is good enough for @ldts's purposes the other changes can wait a little. I can propose something.

Reply:
@jenswi-linaro, looks you're right, no need for specific finalcall stuff.
But then I wonder: @ldts, can't you use existing `driver_init_late()` straight?

Reply:
@etienne-lms  The timing requirement -  to my particular i2c use case ior crypto - is given by whenever the last OP-TEE crypto operation before booting the normal world is executed: at that point is when we need to switch I2C drivers).

But yes I did test with driver_init_late and that worked as well. However I think that having a call that signifies  that  the normal world is booting would be nicer (ie, the semantic distance between the actions being taken in the initcall and the reasons would be clear for everyone to see......not sure if you know what I mean?)

Reply:
> (ie, the semantic distance between the actions being taken in the initcall and the reasons would be clear for everyone to see......not sure if you know what I mean?)

I get you point. Using an explicit `final_xxx` looks better.

That's said, I still prefer `final_init` or `optee_boot_final` or like over `early_boot_final`.

Reply:
I've done some refactoring in https://github.com/OP-TEE/optee_os/pull/3919

Reply:
ah cool. will use those instead. thanks!